### PR TITLE
Add a few functional test covering file globbing in Project

### PR DESCRIPTION
**Pull request by troydai**
Set up a baseline to prepare for integrating new file globbing utilities in Project.

### CODE REVIEW

File: test/Microsoft.Framework.Runtime.FunctionalTests/Utilities/AssertHelpers.cs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        public static void SortAndEqual<T>(IEnumerable<T> expect, IEnumerable<T> actual, IComparer<T> comparer)
+        {
+            var expectArray = expect.OrderBy(elem => elem, comparer).ToArray();
+            var actualArray = actual.OrderBy(elem => elem, comparer).ToArray();
+
+            var missing = new List<T>();
+            var extra = new List<T>();
+
+            int i = 0, j = 0;
+            while (i < expectArray.Length && j < actualArray.Length)
```
Comment (ChengTian) [inline test/Microsoft.Framework.Runtime.FunctionalTests/Utilities/AssertHelpers.cs]: Too complicated. How about the following approach:
```c#
var missing = expect.Except(actual);
var extra = actual.Except(expect);
if (!missing.Any() && !extra.Any())
{
  // They are equal
}
```

File: test/Microsoft.Framework.Runtime.FunctionalTests/ProjectFileGlobbing/ProjectFileGlobbingFacts.cs [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+                "src/project2/content1.txt",
+                "src/project2/obj/object.o",
+                "src/project2/bin/object",
+                "lib/source6.cs",
+                "lib/sub3/source7.cs",
+                "lib/sub4/source8.cs",
+                "res/resource1.text",
+                "res/resource2.text",
+                "res/resource3.text");
+
```
Comment (ChengTian) [inline test/Microsoft.Framework.Runtime.FunctionalTests/ProjectFileGlobbing/ProjectFileGlobbingFacts.cs]: Looks like we are reinventing some functional test utils here: https://github.com/aspnet/XRE/tree/dev/test/Microsoft.Framework.PackageManager.FunctionalTests

@davidfowl , we need to decide the DSL used to describe directory structures. Do we want the plain list above, or we want a single JSON string like:
```c#
var expectedOutputDir = @"{
  '.': ['project.json', 'Config.json', 'Program.cs', 'build_config1.bconfig'],
  'Views': {
    'Home': ['index.cshtml'],
    'Shared': ['_Layout.cshtml']
  },
  'Controllers': ['HomeController.cs'],
  'Models': ['User.cs', 'build_config2.bconfig'],
  'Build': ['build_config3.bconfig'],
  'packages': {}
}";
``` 
Plain list is straightforward but has some redundancy. JSON describes the dir tree structure in a intuitive way but might need more effort to understand.

File: test/Microsoft.Framework.Runtime.FunctionalTests/Utilities/AssertHelpers.cs (hunk 1 above)
Comment (troydai) [inline test/Microsoft.Framework.Runtime.FunctionalTests/Utilities/AssertHelpers.cs]: Good point ...

File: test/Microsoft.Framework.Runtime.FunctionalTests/ProjectFileGlobbing/ProjectFileGlobbingFacts.cs (hunk 2 above)
Comment (troydai) [inline test/Microsoft.Framework.Runtime.FunctionalTests/ProjectFileGlobbing/ProjectFileGlobbingFacts.cs]: Do we have any existing code layout the base for a JSON descriptive test files generator already?

Comment (ChengTian) [inline test/Microsoft.Framework.Runtime.FunctionalTests/ProjectFileGlobbing/ProjectFileGlobbingFacts.cs]: Yes, please check functional tests for `kpm bundle`: https://github.com/aspnet/XRE/blob/dev/test/Microsoft.Framework.PackageManager.FunctionalTests/KpmBundleTests.cs#L49

Comment (davidfowl) [inline test/Microsoft.Framework.Runtime.FunctionalTests/ProjectFileGlobbing/ProjectFileGlobbingFacts.cs]: I kinda prefer the flat list as it's clearer and easier to understand. @ChengTian is there anything we can't represent this way?

File: test/Microsoft.Framework.Runtime.FunctionalTests/Utilities/AssertHelpers.cs (hunk 1 above)
Comment (troydai) [inline test/Microsoft.Framework.Runtime.FunctionalTests/Utilities/AssertHelpers.cs]: It is slow, O(n^2) maybe. Mind is O(n).

But I guess performance is not important here ...

Comment (ChengTian) [inline test/Microsoft.Framework.Runtime.FunctionalTests/Utilities/AssertHelpers.cs]: Yeah, it's not a performance-critical part. We prefer readability over performance.

File: test/Microsoft.Framework.Runtime.FunctionalTests/ProjectFileGlobbing/ProjectFileGlobbingFacts.cs (hunk 2 above)
Comment (ChengTian) [inline test/Microsoft.Framework.Runtime.FunctionalTests/ProjectFileGlobbing/ProjectFileGlobbingFacts.cs]: If we represent empty folder as
```c#
context.AddFiles("src/empty/")
```
then flatten list is as expressive as JSON dir tree.

File: test/Microsoft.Framework.Runtime.FunctionalTests/Utilities/AssertHelpers.cs (hunk 1 above)
Comment (davidfowl) [inline test/Microsoft.Framework.Runtime.FunctionalTests/Utilities/AssertHelpers.cs]: I'm not sure why you want to optimize the tests. Is it really just doing sort and Equal?

```C#
Enumerable.Equals(collection1.OrderBy(...), collection2.OrderBy(...))
```

File: test/Microsoft.Framework.Runtime.FunctionalTests/ProjectFileGlobbing/ProjectFileGlobbingFacts.cs (hunk 2 above)
Comment (ChengTian) [inline test/Microsoft.Framework.Runtime.FunctionalTests/ProjectFileGlobbing/ProjectFileGlobbingFacts.cs]: @troydai , can you add
`public static DirTree CreateFromList(params string[] fileRelativePaths)`
to
`Microsoft.Framework.FunctionalTestUtils.DirTree`
and use `DirTree` class? `DirTree` already does the work you need here and it also contains some other handy utils.

File: test/Microsoft.Framework.Runtime.FunctionalTests/Utilities/AssertHelpers.cs (hunk 1 above)
Comment (ChengTian) [inline test/Microsoft.Framework.Runtime.FunctionalTests/Utilities/AssertHelpers.cs]: A little bit more than that, we want to print out missing items and extra items when they are not equal. It's very helpful when you want to investigate a failed test case.

File: test/Microsoft.Framework.Runtime.FunctionalTests/ProjectFileGlobbing/ProjectFileGlobbingFacts.cs (hunk 2 above)
Comment (davidfowl) [inline test/Microsoft.Framework.Runtime.FunctionalTests/ProjectFileGlobbing/ProjectFileGlobbingFacts.cs]: We should (in another change) move away from the json format and move to the flat list.

File: test/Microsoft.Framework.Runtime.FunctionalTests/Utilities/AssertHelpers.cs (hunk 1 above)
Comment (davidfowl) [inline test/Microsoft.Framework.Runtime.FunctionalTests/Utilities/AssertHelpers.cs]: More linq?

Comment (troydai) [inline test/Microsoft.Framework.Runtime.FunctionalTests/Utilities/AssertHelpers.cs]: @ChengTian 's solution is enough.

Comment (davidfowl) [inline test/Microsoft.Framework.Runtime.FunctionalTests/Utilities/AssertHelpers.cs]: Exactly, more linq

**Comment by troydai** [comment]
@davidfowl @ChengTian 

Looks good now?

**Comment by davidfowl** [comment]
@troydai I think you should add test cases based on what exists today in all of our repos. Paths with slash in the other direction, absolute paths, not sure what else we do. EF has some interesting ones.

File: test/Microsoft.Framework.Runtime.FunctionalTests/ProjectFileGlobbing/ProjectFileGlobbingFacts.cs (hunk 2 above)
Comment (ChengTian) [inline test/Microsoft.Framework.Runtime.FunctionalTests/ProjectFileGlobbing/ProjectFileGlobbingFacts.cs]: Filed an issue: https://github.com/aspnet/XRE/issues/1184

**Comment by troydai** [comment]
Added a dozen new tests and find a bug in current runtime: https://github.com/aspnet/XRE/issues/1187

By searching the `project.json` samples in the `Universe` following three patterns are additionally used:
1. `**\\*.cs`
2. `..\\<upperlevel_folder>\\*.cs`
3. `..\\<upperlevel_folder>\\<literal_filename>`

File: test/Microsoft.Framework.Runtime.FunctionalTests/ProjectFileGlobbing/ProjectFileGlobbingFacts.cs [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+            VerifyFilePathsCollection(testProject.SourceFiles,
+                @"src\project\sub\source2.cs", 
+                @"src\project\sub\source3.cs");
+        }
+
+        [Fact]
+        public void IncludeCodeAbsoluePath()
+        {
+            var absolutPath = Path.Combine(Context.RootPath, @"src\project2\sub2\source5.cs");
```
Comment (davidfowl) [inline test/Microsoft.Framework.Runtime.FunctionalTests/ProjectFileGlobbing/ProjectFileGlobbingFacts.cs]: thats not how you spell absolute

File: test/Microsoft.Framework.Runtime.FunctionalTests/ProjectFileGlobbing/ProjectFileGlobbingFacts.cs [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+}
+", @"src\project");
+
+            VerifyFilePathsCollection(testProject.SourceFiles,
+                @"src\project\sub\source2.cs", 
+                @"src\project\sub\source3.cs");
+        }
+
+        [Fact]
+        public void IncludeCodeAbsoluePath()
```
Comment (davidfowl) [inline test/Microsoft.Framework.Runtime.FunctionalTests/ProjectFileGlobbing/ProjectFileGlobbingFacts.cs]: Neither is this :smile: :

**Comment by troydai** [comment]
@davidfowl take another look?

**Comment by ChengTian** [comment]
When doing `kpm bundle`, we want to exclude folders starting with a dot, for example `.git` folder.
Check `Project._defaultBundleExcludePatterns`, which contains `@"**\.*\**"`. We need a test for this pattern.

**Comment by troydai** [comment]
@ChengTian update the test.

**Comment by ChengTian** [comment]
:shipit:

**Pull request merged**